### PR TITLE
Implement weight as part of the allreduce protocol, not matchmaking

### PR DESCRIPTION
**Pull request by justheuristic**
This PR allows specifying allreduce weights in AllReduceRunner, instead of gathering them during matchmaking.

This will allow peers to use their actual batch size in both DPU and advance matchmaking (aka @yhn112 -style matchmaking)

__[WIP]__ implement advance matchmaking as a working example

### CODE REVIEW

File: hivemind/averaging/allreduce.py [hunk 1]
```diff
@@ -37,13 +38,12 @@ class AllReduceRunner(ServicerBase):
     :param prefix: namespace for servicer's RPCs (typically, equal to prefix for group keys)
     :param group_id: unique identifier of this specific all-reduce run
     :param tensors: local tensors that should be averaged with groupmates
-    :param tensors: local tensors that should be averaged with groupmates
```
Comment (justheuristic) [inline hivemind/averaging/allreduce.py:40]: yeah, that was in the master code :)

File: hivemind/averaging/allreduce.py [hunk 2]
```diff
@@ -180,9 +182,10 @@ async def _generate_input_for_peer(self, peer_index: int) -> AsyncIterator[avera
             code=averaging_pb2.PART_FOR_AVERAGING,
             group_id=self.group_id,
             tensor_part=first_part,
+            metadata=self._weight_binary,
```
Comment (borzunov) [inline hivemind/averaging/allreduce.py:185]: Can we create a new `double weight;` field instead and avoid encoding the weight to the binary format manually?

**Review by borzunov — CHANGES_REQUESTED** [review]
I've left one comment, everything else is good :)

File: hivemind/averaging/allreduce.py (hunk 2 above)
Comment (justheuristic) [reply hivemind/averaging/allreduce.py]: done so

File: hivemind/proto/averaging.proto [hunk 3]
```diff
@@ -45,7 +45,7 @@ message AveragingData {
   bytes group_id = 2;       // a unique group identifier, same as in MessageFromLeader
   bytes peer_id = 3;        // sender's rpc peer_id, used for coordination
   Tensor tensor_part = 4;   // either peer's local tensor part (rpc input) or group average of this part (rpc output)
-  bytes metadata = 5;       // reserved user-extendable metadata
+  double weight = 5;         // peers will be averaged in proportion to these weights
```
Comment (borzunov) [inline hivemind/proto/averaging.proto:48]:
```suggestion
  double weight = 5;        // peers will be averaged in proportion to these weights
```

nit: There are too many spaces currently :)

**Review by borzunov — APPROVED** [review]

**Pull request merged**